Added torch.linalg.tensorsolve

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.tensorsolve` function that matches `numpy.linalg.tensorsolve`.

Ref #42666.

IvanYashchuk committed:
Added torch.linalg.tensorsolve

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8319,6 +8319,10 @@
   python_module: linalg
   variants: function
 
+- func: linalg_tensorsolve(Tensor self, Tensor input2, int[]? axes=None) -> Tensor
```
"input2" -> "other" to be consistent with other PyTorch functions.

"axes" -> "dims" to be consistent with other PyTorch functions. Note that PyTorch's bindings should remap "axes->dims", too.

mruberry replied:
What's the plan for out support?

IvanYashchuk replied:
I wasn't thinking about out support, since NumPy doesn't do that. I guess it can be added to this PR.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
```
Our test suite now requires NumPy so we don't need these skips anymore

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    @dtypesIfCPU(torch.float, torch.double, torch.cfloat, torch.cdouble)
```
This doesn't need a CPU dtype override since the dtypes are the same as the default dtypes (above)

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
+
+    # TODO: once "solve_cuda" supports complex dtypes, they shall be added to above test
+    @unittest.expectedFailure
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Can remove this

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
+
+        a_shapes = [(2, 3, 6), (3, 4, 4, 3)]
+        axes = [None, (0, 2)]
+        for a_shape, ax in itertools.product(a_shapes, axes):
```
Nice enumeration over the product space here.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
+        def run_test(a_shape, axes):
+            a = torch.rand(a_shape, dtype=dtype, device=device)
+            b = torch.rand(a_shape[:2], dtype=dtype, device=device)
+            result = torch.linalg.tensorsolve(a, b, axes=axes)
```
Update to "dims=" (see above note)

mruberry commented on torch/csrc/api/include/torch/linalg.h:
```diff
@@ -53,4 +57,18 @@ inline Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string o
   return detail::norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+/// Solving of the tensor equation `a x = b`
```
"Solving of the tensor equation" -> "Solves the equation"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(a, b, axes=None) -> Tensor
```
tensorsolve(input, other, dims=None)

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(a, b, axes=None) -> Tensor
+
+Computes the solution to the tensor equation ``a x = b`` for ``x``.
```
"Computes a tensor \`x\` such that \`\`input @ x = other\`\`."

I think that will render correctly?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+
+Computes the solution to the tensor equation ``a x = b`` for ``x``.
+
+Suppose that `:attr:`b` is equivalent to ``torch.tensordot(a, x)``.
```
If the initial sentence is updated I don't think this sentence is needed?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+    a (Tensor): left-hand-side tensor of size ``b.shape + Q``.
+        ``Q`` is the shape of that sub-tensor of `:attr:`a` consisting of the appropriate
+        number of its rightmost indices, and must be such that ``prod(Q) == prod(b.shape)``.
+    b (Tensor): right-hand-side tensor of any size.
```
"solution to tensordot(:attr:\`input\`, x). May be any size."

Is that true that it may actually be any size? Will an empty tensor work, for example?

IvanYashchuk replied:
Size == shape here.

It's true that can be any size. Choosing `b` (or `other`) puts restrictions on shape of `a` (or `input`).
An empty tensor would work if `lda` parameter (4th argument) in `lapackSolve` call was handled correctly.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+This function computes tensor ``x`` from `:attr:`a` and `:attr:`b`.
+
+Args:
+    a (Tensor): left-hand-side tensor of size ``b.shape + Q``.
```
I'm not sure what this is trying to say. I like that it's trying to describe the input shape but I'm not following the description.

IvanYashchuk replied:
Maybe the confusion comes from using the word "size", I don't know what is the preferred term now, but I was referring to the shape of the tensor here (`tensor.size() == tensor.shape`).
`other.shape` is a tuple in NumPy and `torch.Size` in PyTorch and `Q` is a tuple, then `input.shape == other.shape + Q`.
I should change `Q`->`x.shape`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+        ``Q`` is the shape of that sub-tensor of `:attr:`a` consisting of the appropriate
+        number of its rightmost indices, and must be such that ``prod(Q) == prod(b.shape)``.
+    b (Tensor): right-hand-side tensor of any size.
+    axes (Tuple[int]): axes (dimensions) of `:attr:`a` to reorder to the right, before solve.
```
The use of the dims argument will have to be mentioned in the description of the function above, too. "reorder" is an imprecise term that will need careful clarification.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+    axes (Tuple[int]): axes (dimensions) of `:attr:`a` to reorder to the right, before solve.
+        If None (default), no reordering is done.
+
+    Returns:
```
I don't think this needs a returns section.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+    >>> x = torch.linalg.tensorsolve(a, b)
+    >>> x.shape
+    torch.Size([2, 3, 4])
+    >>> torch.allclose(torch.tensordot(a, x, dims=3), b)
```
It's important to relate the dims argument here to the dims argument to this function.

IvanYashchuk replied:
They are not related. `dims` argument to the current function tells how to reorder the dimensions of input.
I've changed the documentation to indicate that `dims` argument to `tensordot` should be `len(x.shape)` or `x.ndim`.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
         self.assertRaisesRegex(RuntimeError, "duplicate or invalid", torch.norm, x, "nuc", (0, 0))
         self.assertRaisesRegex(IndexError, "Dimension out of range", torch.norm, x, "nuc", (0, 2))
 
+    @skipCUDAIfNoMagma
```
Update common_methods_invocations.py's method_tests, too.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,33 @@
+
+Suppose that `:attr:`b` is equivalent to ``torch.tensordot(a, x)``.
+This function computes tensor ``x`` from `:attr:`a` and `:attr:`b`.
+
```
Add a brief "Supports..." statement here.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,33 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return linalg_norm_out_impl(result, self, c10::nullopt, ord, opt_dim, keepdim, opt_dtype);
 }
 
+Tensor linalg_tensorsolve(const Tensor& self, const Tensor& input2, optional<IntArrayRef> axes) {
```
Add a brief comment describing the approach

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
+            expected = np.linalg.tensorsolve(a.cpu().numpy(), b.cpu().numpy(), axes=axes)
+            self.assertEqual(result, expected)
+
+        a_shapes = [(2, 3, 6), (3, 4, 4, 3)]
```
Are there odd/extremal edge cases to test, too, like how empty tensors (on both the lhs and rhs) are handled?

mruberry reviewed (changes requested):
Hey @IvanYashchuk!

This is a good PR that's already near completion. It needs some consistency updates (arg names, out support), maybe a few more test cases, and some additional doc clarity. As I'm sure @kurtamohler or @peterbell10 will tell you I'm a stickler for super-precise clear documentation. Writing documentation is hard and the current docs aren't bad, but I think it's important we spend the time to make our documentation great. It is, after all, how many PyTorch users learn how to use the framework.

IvanYashchuk committed:
Removed unnecessary @dtypesIfCPU

IvanYashchuk committed:
Removed unnecessary skip for NumPy

IvanYashchuk committed:
Added a comment on the approach

IvanYashchuk committed:
Renamed axes -> dims, input2 -> other

IvanYashchuk committed:
Fixed lda/ldb parameters for lapackSolve/magmaSolve

IvanYashchuk committed:
Changed reshape not to use -1

IvanYashchuk committed:
Updated documentation

IvanYashchuk committed:
Added test for empty inputs

IvanYashchuk commented:
In order to handle empty input cases correctly I fixed an error related to `lda` parameter in `apply_solve` in `BatchLinearAlgebra.{cpp, cu}`

IvanYashchuk committed:
Removed unused import

IvanYashchuk committed:
Use at::movedim instead of my implementation

IvanYashchuk committed:
Removed 'note' section

IvanYashchuk committed:
Moved the empty test case to a separate test

IvanYashchuk committed:
Added a check that the input is reshapeable to square matrix

IvanYashchuk committed:
Added out= variant

IvanYashchuk committed:
Added entry to the common_methods_invocations.py

IvanYashchuk committed:
Added entry to the overrides.py

IvanYashchuk commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1601,6 +1601,33 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+      auto it = std::find(permutation_axes.begin(), permutation_axes.end(), k);
+      std::rotate(it, it + 1, permutation_axes.end());
+    }
+    self_ = self_.permute(permutation_axes);
```
Thank you! I was not aware of this function.

IvanYashchuk committed:
Reorder dimensions -> move dimensions

IvanYashchuk committed:
Added out=None to the function signature in the docs

IvanYashchuk committed:
Updated documentation. Added an example with dims argument

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Fixed typo

IvanYashchuk committed:
Corrected example in the docs

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -921,6 +922,69 @@ def test_nuclear_norm_exceptions_old(self, device):
+    @dtypesIfCUDA(torch.float, torch.double)
+    def test_tensorsolve_empty(self, device, dtype):
+        # Check for empty inputs. NumPy does not work for these cases.
+        a = torch.empty(0, 0, 1, 2, 3, 0)
```
```suggestion
        a = torch.empty(0, 0, 1, 2, 3, 0, dtype=dtype, device=device)
```

IvanYashchuk committed:
Corrected test_tensorsolve_empty

IvanYashchuk committed:
Corrected test_tensorsolve_empty (again)

IvanYashchuk commented on torch/csrc/api/include/torch/linalg.h:
```diff
@@ -28,6 +28,10 @@ inline Tensor& norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return torch::linalg_norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+inline Tensor tensorsolve(const Tensor& self, const Tensor& other, optional<IntArrayRef> dims) {
```
- [ ] Add also `tensorsolve_out` here.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Added tensorsolve_out to linalg.h

IvanYashchuk committed:
Move tensorsolve to the end of the list

IvanYashchuk committed:
Added test_tensorsolve_non_contiguous

IvanYashchuk committed:
Changed rand -> randn

IvanYashchuk commented:
@mruberry, I've added a few changes to this PR. I think it's in better shape now.
I updated the documentation, I hope now it's more clear even though that's a complicated function.
`out=` support was added as well. I updated the tests to include empty, non-contiguous inputs and error checks for invalid input.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+
+  // Check whether the self tensor can be reshaped to the 2D square matrix
+  TORCH_CHECK(result_product == other_product,
+    "Expected self to be a 'square' tensor such that prod(result.shape) == prod(other.shape), but got ",
```
This is a little wonky since "result" is not user-visible and it would be more accurate to say, as the comment does, that self can be reshaped into a `square` tensor for purposes of this problem, not that it is square. I think taking the time to add more detail here and relate the requirement back to self and other would be helpful.

IvanYashchuk replied:
I've changed the message to `Expected self to satisfy the requirement prod(self.shape[other.ndim:]) == prod(self.shape[:other.ndim]), but got `

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+    self_ = at::movedim(self_, dims.value(), dest_axes);
+  }
+
+  auto expected_result_shape = self_.sizes().slice(other.dim(), self.dim() - other.dim());
```
This is OK. This check could also occur after the computation by verifying that result_tmp and result have the same shape.

IvanYashchuk replied:
For large inputs the computation can take a long time. I thought it's a good idea to exit the function before doing actual computations if those will be wasteful.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+
+  // We've already reordered self with dims, let's pass that result
+  Tensor result_tmp = at::linalg_tensorsolve(self_, other, c10::nullopt);
+  result.copy_(result_tmp);
```
An interesting side affect of this approach is that it will allow copying to a result tensor with a different dtype than result_tmp. This pattern of computing a result and then copying to an out tensor is common enough we should consider making a helper function for it.

IvanYashchuk replied:
I don't know what's the expected behavior in general, so there is a check now that doesn't allow different dtypes.
https://github.com/pytorch/pytorch/pull/46142/files/497bc2241ea7ceae548277aac989849ad5a1bb0b#diff-db2b91c51cffd1fba933933ac20de256f5f0ced095890a1a49a17d73037001a2R1648

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
+The resulting tensor ``x`` has the shape equal to ``Q``, where ``Q`` is the shape
```
"has the shape equal to Q" -> "has the same shape as Q"

mruberry replied:
Before describing the shape of a result, I think this should start a new paragraph and discuss the shape requirement for input and other. That should make defining the shape of x easier. 

As I understand it, if the shape of input is (a, b, c, d, ... z), then the shape of other must (a, b, c, d...) and the shape of x will be (..., x, y z). Is that correct?

IvanYashchuk replied:
> As I understand it, if the shape of input is (a, b, c, d, ... z), then the shape of other must (a, b, c, d...) and the shape of x will be (..., x, y z). Is that correct?

That is correct, but ...
Here is how I understand the current NumPy doc. First, the `other` tensor is fixed, that's why it can be of any shape. Then based on `other.shape` the requirements on `input` are posed. If `other.shape` is `(a, b, c, d)` then:
- `input.shape` must start with `a, b, c, d`
  - at this stage `torch.movedim(input, dims, (-(len(dims)+1), ..., -2, -1))` can be used to match the requirement.
 (moving dimensions specified in `dims` to the right end keeping the order)
- `input.shape = (a, b, c, d) + (..., x, y, z) = (a, b, c, d, ..., x, y, z)`
- `x.shape = (..., x, y, z)`

For the function to work `input` must satisfy the requirement `prod(..., x, y, z)==prod(a, b, c, d)`.

Alternatively, we can start with `input` and `other.ndim`, then:
- `other.shape` must be such that `other.shape == input.shape[0:other.ndim]` (that point is a bit problematic since we pose the requirement on `other` using `other.ndim`.
- `x.shape = input.shape[other.ndim:]`

`input` must satisfy the requirement `prod(input.shape[other.ndim:] == input.shape[:other.ndim]`.

I think that the second variant is better, because then we don't introduce any `Q`, and the "squareness" requirement is directly relatable to `tensorinv` function with `ind = other.ndim`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
```
Great first sentence.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+Before computing ``x``, dimensions of `:attr:`input` can be moved to the right
+using :attr:`dims` to match the shape requirements.
+
+Supports real and, only on the CPU, complex inputs.
```
"Supports real" -> "Supports float"

IvanYashchuk replied:
Both real and complex numbers are floating-point numbers. What about 
```suggestion
Supports real-valued and, only on the CPU, complex-valued inputs.
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+
+Supports real and, only on the CPU, complex inputs.
+
+.. note:: If :attr:`input` is not a 'square' tensor, meaning it does not satisfy the requirement
```
This note doesn't seem correct since it doesn't consider the dims argument, for example. Also, I think we should avoid describing the input as "square," since square suggests that input is a nxn matrix.

IvanYashchuk replied:
I've changed this note to 
```
.. note:: If :attr:`input` does not satisfy the requirement
          ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``
          after (optionally) moving the dimensions using using :attr:`dims`, then a RuntimeError will be thrown.
```

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+    input (Tensor): left-hand-side tensor of shape ``other.shape + Q``.
+        ``Q`` is the shape of that sub-tensor of `:attr:`input` consisting of the appropriate
+        number of its rightmost dimensions, and must be such that ``prod(Q) == prod(other.shape)``.
+    other (Tensor): right-hand-side tensor of any shape.
```
Isn't this constrained just as input is above?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,51 @@
+        ``Q`` is the shape of that sub-tensor of `:attr:`input` consisting of the appropriate
+        number of its rightmost dimensions, and must be such that ``prod(Q) == prod(other.shape)``.
+    other (Tensor): right-hand-side tensor of any shape.
+    dims (Tuple[int]): dimensions of `:attr:`input` to be moved to the right, before solve.
```
Saying the dimensions are "moved to right" is a little vague (how far to the right? in what order? what happens to the other dims?). Can this describe the behavior in terms of another function, like movedim? Is it, for example, that each dim is "movedim'd" to the -1 position in the order it appears it the tuple?

IvanYashchuk replied:
How about this?
```
    dims (Tuple[int]): dimensions of `:attr:`input` to be moved to the end keeping the order,
                       i.e. ``movedim(input, dims, range(len(dims) - input.ndim, 0)), before solve.
                       If None (default), no moving of dimensions is done.
```

mruberry commented:
Hey Ivan, 

Overall this PR and its latest updates look great. This is a tricky function to describe, however. I made a few more comments/suggestions about error messages and the docs.

IvanYashchuk committed:
Changed the error message about shape requirements for the input

IvanYashchuk committed:
Updated the documentation, trying to make it more clear

IvanYashchuk commented:
So JIT started to fail, need to fix that.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Added dispatch: Math

IvanYashchuk committed:
Use at::native::resize_output in linalg_tensorsolve_out

IvanYashchuk committed:
Updated errors and warnings tests

IvanYashchuk committed:
Added non-contiguous out test case

IvanYashchuk committed:
Skip jit annotation testing with exclude_tensor_method

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

IvanYashchuk committed:
Fix tests

IvanYashchuk committed:
Fixed typos in docs

IvanYashchuk commented:
@mruberry, this is ready for another round.
I've updated the general description of the function with the second option from here https://github.com/pytorch/pytorch/pull/46142#discussion_r510054412
`dims` argument is described as here https://github.com/pytorch/pytorch/pull/46142#discussion_r510072369

Updated `linalg_tensorsolve_out` to use `at::native::resize_output`.
Explicit `dispatch: Math` is now used `native_functions.yml`.
Added non-contiguous out= tests cases.

mruberry commented on torch/csrc/api/include/torch/linalg.h:
```diff
@@ -53,4 +61,22 @@ inline Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string o
   return detail::norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+/// Solves the tensor equation `a x = b`
```
This line needs an update to reflect the updates in the regular documentation.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
+The resulting tensor ``x`` has the shape equal to ``input[other.ndim:]``.
```
"the shape equal to" -> "the same shape as"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
+The resulting tensor ``x`` has the shape equal to ``input[other.ndim:]``.
+Before computing ``x``, dimensions of :attr:`input` can be moved using :attr:`dims` to match the shape requirements.
```
Since the shape requirements haven't been introduced yet and "moved" is not well-defined, I suggest removing this sentence and leaving the documentation of dims to the more detailed note below.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+
+.. note:: If :attr:`input` does not satisfy the requirement
+          ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``
+          after (optionally) moving the dimensions using using :attr:`dims`, then a RuntimeError will be thrown.
```
"using using :attr:\`dims\`" -> "using :attr:\`dims\` (see below for details)"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+          after (optionally) moving the dimensions using using :attr:`dims`, then a RuntimeError will be thrown.
+
+Args:
+    input (Tensor): left-hand-side tensor, it must satisfy the requirement
```
Put quotation marks around "left-hand-side" so it's clear that's an informal term.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+Args:
+    input (Tensor): left-hand-side tensor, it must satisfy the requirement
+                    ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``.
+    other (Tensor): right-hand-side tensor of shape ``input.shape[other.ndim]``.
```
Quotes around "right-hand-side" here, too

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -140,3 +140,49 @@
+    input (Tensor): left-hand-side tensor, it must satisfy the requirement
+                    ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``.
+    other (Tensor): right-hand-side tensor of shape ``input.shape[other.ndim]``.
+    dims (Tuple[int]): dimensions of :attr:`input` to be moved to the end keeping the order,
```
"to be moved to the end keeping the order, ... " -> "to be moved before the computation. Equivalent to calling \`\`input = movedim(input, dims, range(len(dims) - input.ndim, 0))\`\`. If None (default), no dimensions are moved."

mruberry reviewed (approved):
Great work, @IvanYashchuk. I just made a few comments/suggestions that should be easy to implement. Let me know when this is ready to merge!

IvanYashchuk committed:
Changed 'the shape equal to' -> 'the same shape as'

IvanYashchuk committed:
Remove repeating word

IvanYashchuk committed:
Use quotation marks

IvanYashchuk committed:
Updated 'dims' description

IvanYashchuk committed:
Updated doc in linalg.h

IvanYashchuk committed:
Removed dims sentence from the main description

IvanYashchuk commented:
Thank you for reviewing this pull request and for all the suggestions, @mruberry!
Now this is ready to merge 🎉

mruberry commented:
Hey @IvanYashchuk, would you rebase this, please? The internal build tools are hitting an issue that I expect a rebase to solve.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into numpy-tensorsolve

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1444,6 +1444,8 @@ def method_tests():
         ('__getitem__', torch.randn(S, S, S), (dont_convert([[0, 2, 3], [1, 3, 3],
                                                              torch.LongTensor([0, 0, 2])]),), 'adv_index_var'),
         ('to_sparse', (S, S), (), '', (), (), [], lambda x: x.to_dense()),
+        ('linalg.tensorsolve', (2, S, M), ((2, S),),
```
Most of our internal build issues have been resolved but there's still an internal problem resolving the references to the tests generated here. Would you just revert the changes to common_methos_invocations.py for a moment? We should move these tests to use OpInfos soon, anyway. The temporary gap in test coverage is OK. 

Sorry to keep bringing these internal tooling issues to you, @IvanYashchuk.

IvanYashchuk replied:
Okay, I removed it.

IvanYashchuk committed:
Revert 1399edf; removed entry from common_methods_invocations.py
